### PR TITLE
- fixes ResonanceMolSupplier bug in perceiving conjugated groups

### PR DESCRIPTION
**Pull request by ptosco**
#### Reference Issue
I found a bug where a quinoline is perceived as two disconnected conjugated groups whereas obviously it should be a single, connected conjugated group. The bug is triggered by atom/bond ordering, and is caused by the conjugated group assignment not being robust against different ordering of bonds in the molecule.

### CODE REVIEW

**Comment by ptosco** [comment]
@greglandrum not sure why the Windows Azure pipeline failed. I don't think the error has anything to do with the PR. FWIW, the PR works OK on my Windows machine.

**Comment by greglandrum** [comment]
It looks like something is horked on Azure DevOps side.
I will see if I can figure that out

File: Code/GraphMol/resMolSupplierTest.cpp [hunk 1]
```diff
@@ -713,6 +713,95 @@ void testCrambin() {
   delete crambin;
 }
 
+void testConjGrpPerception() {
+  BOOST_LOG(rdInfoLog) << "-----------------------\n"
+                       << "testConjGrpPerception" << std::endl;
+  RWMol *mol1 = MolBlockToMol(
```
Comment (greglandrum) [inline Code/GraphMol/resMolSupplierTest.cpp]: In modern C++ we can use raw string literals and no longer have to do these horrible multi-line strings.
Take a look at https://github.com/rdkit/rdkit/blob/master/Code/GraphMol/FileParsers/testMolSupplier.cpp#L2375

File: Code/GraphMol/Resonance.cpp [hunk 2]
```diff
@@ -1189,26 +1189,37 @@ void ResonanceMolSupplier::mainLoop(unsigned int ti, unsigned int nt) {
...
-      for (unsigned int j = 0; (d_bondConjGrpIdx[i] == -1) && (j < nb); ++j) {
-        if ((i == j) || (d_bondConjGrpIdx[j] == -1)) continue;
-        const Bond *bj = d_mol->getBondWithIdx(j);
-        if ((bj->getBeginAtomIdx() == biBeginIdx) ||
-            (bj->getBeginAtomIdx() == biEndIdx) ||
-            (bj->getEndAtomIdx() == biBeginIdx) ||
-            (bj->getEndAtomIdx() == biEndIdx))
-          d_bondConjGrpIdx[i] = d_bondConjGrpIdx[j];
+  for (d_nConjGrp = 0; true; ++d_nConjGrp) {
+    for (unsigned int i = 0; i < nb && bondIndexStack.empty(); ++i) {
```
Comment (greglandrum) [inline Code/GraphMol/Resonance.cpp]: Seems like this loop would be clearer as:
```suggestion
    for (const auto bond : d_mol->bonds() ) {
      if(bond->getIsConjugated() && d_bondConjGrpIdx[bond->getIdx()]==-1){
          bondIndexStack.push(bond->getIdx());
          break;
       }
    }
```

File: Code/GraphMol/Resonance.cpp [hunk 3]
```diff
@@ -1189,26 +1189,37 @@ void ResonanceMolSupplier::mainLoop(unsigned int ti, unsigned int nt) {
...
+    if (bondIndexStack.empty()) break;
+    while (!bondIndexStack.empty()) {
+      unsigned int i = bondIndexStack.top();
+      bondIndexStack.pop();
+      const Bond *bi = d_mol->getBondWithIdx(i);
+      for (const Atom *bondAtom: { bi->getBeginAtom(), bi->getEndAtom() }) {
+        // loop over neighbors of the bondAtom
+        unsigned int aiSelf = bondAtom->getIdx();
+        ROMol::ADJ_ITER nbrIdx, endNbrs;
+        boost::tie(nbrIdx, endNbrs) = d_mol->getAtomNeighbors(bondAtom);
```
Comment (greglandrum) [inline Code/GraphMol/Resonance.cpp]: Seems simpler to just use `d_mol->getAtomBonds()` here.

File: Code/GraphMol/resMolSupplierTest.cpp (hunk 1 above)
Comment (ptosco) [reply Code/GraphMol/resMolSupplierTest.cpp]: @greglandrum Done.

File: Code/GraphMol/Resonance.cpp (hunk 2 above)
Comment (ptosco) [reply Code/GraphMol/Resonance.cpp]: @greglandrum Done.

File: Code/GraphMol/Resonance.cpp (hunk 3 above)
Comment (ptosco) [reply Code/GraphMol/Resonance.cpp]: @greglandrum Done.

**Comment by ptosco** [comment]
@greglandrum All done.

**Pull request merged**